nomos-services/membership/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+    B::Settings: Clone,
+{
+    type Settings = BackendSettings<B::Settings>;
+    type State = NoState<Self::Settings>;
+    type StateOperator = NoOperator<Self::State>;
+    type Message = MembershipMessage;
+}
+
+#[async_trait]
+impl<B, S, RuntimeServiceId> ServiceCore<RuntimeServiceId>
+    for MembershipService<B, S, RuntimeServiceId>
+where
+    B: MembershipBackend + Send + Sync + 'static,
+    B::Settings: Clone,
+
+    RuntimeServiceId: AsServiceId<Self>
+        + AsServiceId<S::SdpService>
+        + Clone
+        + Display
+        + Send
+        + Sync
+        + 'static
+        + Debug,
+    S: SdpAdapter + Send + Sync + 'static,
+    <<S as adapters::SdpAdapter>::SdpService as overwatch::services::ServiceData>::Message: 'static,
+{
+    fn init(
+        service_state: OpaqueServiceStateHandle<Self, RuntimeServiceId>,
+        _initstate: Self::State,
+    ) -> Result<Self, overwatch::DynError> {
+        let BackendSettings {
+            backend: backend_settings,
+        } = service_state.settings_reader.get_updated_settings();
+
+        Ok(Self {
+            backend: B::init(backend_settings),
+            service_state,
+            subscribe_txs: HashMap::new(),
+        })
+    }
+
+    async fn run(mut self) -> Result<(), overwatch::DynError> {
+        let sdp_relay = self
+            .service_state
+            .overwatch_handle
+            .relay::<S::SdpService>()
+            .await?;
+
+        let sdp_adapter = S::new(sdp_relay);
+        let mut sdp_stream = sdp_adapter
+            .finalized_blocks_stream()
+            .await
+            .map_err(|e| match e {
+                adapters::SdpAdapterError::Other(error) => error,
+            })?;
+
+        let mut lifecycle_stream = self.service_state.lifecycle_handle.message_stream();
+        loop {
+            tokio::select! {
+                Some(msg) = self.service_state.inbound_relay.recv()  => {
+                    match msg {
+                        MembershipMessage::GetSnapshotAt { reply_channel, block_number, service_type } =>  {
+                            let result = self.backend.get_providers_at(service_type, block_number).await;
+
+                            if let Err(e) = reply_channel.send(result) {
+                                tracing::error!("Failed to send response: {:?}", e);
+                            }
+                        },
+                        MembershipMessage::Subscribe { service_type, result_sender } => {
+                            let tx = if let Some(tx) = self.subscribe_txs.get(&service_type) {
+                                tx.clone()
+                            } else {
+                                let (tx, _) = broadcast::channel(128);
+                                self.subscribe_txs.insert(service_type, tx.clone());
+                                tx
+                            };
+
+                            let stream = make_pin_broadcast_stream(tx.subscribe());
```

Comment:
Shouldn't we write the latest members to the newly created stream after somebody asks for membership?

Reply:
Good point! I will add the latest snapshot to the stream before returning it to the caller.